File: UnityClient/Assets/Scripts/Base/Vector.cs
/*
* Author:  caoshanshan
* Email:   me@dreamyouxi.com

 */
using UnityEngine;
using System.Collections;



public class Vector
{
    public bool Contains(GAObject obj)
    {
        return this.list.Contains(obj) ;
    }
    public void PushBack(GAObject obj)
    {
        this.list.Add(obj);
    }
    public void PushFront(GAObject obj)
    {
        this.list.Insert(0, obj);
    }
    /// <summary>
    /// 移除容器中第一个匹配项
    /// </summary>
    /// <param name="obj"></param>
    public void Remove(GAObject obj)
    {
        this.list.Remove(obj);
    }
    /// <summary>
    /// 移除容器中所有匹配项
    /// </summary>
    /// <param name="obj"></param>
    public void RemoveAll(GAObject obj)
    {
        for (int i = 0; i < list.Count; )
        {
            GAObject b = list[i] as GAObject;
            if (b == obj)
            {
                this.Remove(b);
            }
            else
            {
                ++i;
            }
        }
    }
    /// <summary>
    /// 清空
    /// </summary>
    public void Clear()
    {
        list.Clear();
    }
    public int Count()
    {
        return list.Count;
    }
    public Vector Clone()
    {
        Vector ret = new Vector();
        ret.list = list.Clone() as ArrayList;
        return ret;
    }
    public GAObject this[int index]
    {
        get
        {
            return list[index] as GAObject;
        }

        set
        {
            list[index] = value as GAObject;
        }
    }

    /// <summary>
    /// 相同元素的个数
    /// </summary>
    /// <param name="obj"></param>
    /// <returns></returns>
    public int GetCount(GAObject obj)
    {
        int ret = 0;
        foreach (GAObject obj1 in list)
        {
            if (obj == obj1) ++ret;
        }
        return ret;
    }

    public Vector()
    {
        list.Capacity = 100;
    }

    private ArrayList list = new ArrayList();
}

File: LogicServer/server/scripts/common/xygame/base/vector.lua
--[[

* Author: caoshanshan
* Email: me@dreamyouxi.com
 容器
]]

local t = class("vector");


function t:ctor()
    self.list = { };
    self.size = 0;
end


function t:add(obj)

    table.insert(self.list, obj);
    self.size = self.size + 1;
end

function t:remove(obj)
    if self.size == 0 then return end;

    for k, v in pairs(self.list) do
        if v == obj then
            table.remove(self.list, k);
            self.size = self.size - 1;
            return;
        end
    end
end


function t:clear()
    self.size = 0;
    self.list = { };
end

function t:current_size()
    return self.size;

end

-- cb 返回false 将会终止循环
function t:foreach(cb)
    local dirty_list = { };
    for k, v in pairs(self.list) do
        local ret = cb(k, v);
        if ret == false then
            return;
        end
        if v.is_dirty == true then
            --   table.insert(dirty_list, k);
        end
    end

    for k, v in ipairs(dirty_list) do

        self:remove(v);
    end

end



return t;